### PR TITLE
Ember: delegate `Enumerable` method types to `Array`.

### PR DESCRIPTION
**Pull request by chriskrycho**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    * <https://emberjs.com/api/ember/2.18/classes/Enumerable/methods/filter?anchor=forEach>
    * <https://emberjs.com/api/ember/2.18/classes/Enumerable/methods/filter?anchor=map>
    * <https://emberjs.com/api/ember/2.18/classes/Enumerable/methods/filter?anchor=filter>
    * <https://emberjs.com/api/ember/2.18/classes/Enumerable/methods/filter?anchor=find>
    * <https://emberjs.com/api/ember/2.18/classes/Enumerable/methods/filter?anchor=every>
    * <https://emberjs.com/api/ember/2.18/classes/Enumerable/methods/filter?anchor=reduce>
- ~Increase the version number in the header if appropriate.~
- ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~

Fixes https://github.com/typed-ember/ember-typings/issues/70.

### CODE REVIEW

**Comment by dwickern** [comment]
Can we remove the duplicated definitions in NativeArray?
https://github.com/chriskrycho/DefinitelyTyped/blob/c4173d92cc5e18cc5600d40fed5dbab00bb19201/types/ember/index.d.ts#L1603-L1634

**Comment by chriskrycho** [comment]
Good suggestion, @dwickern. Pushing that up now.

**Review by dwickern — APPROVED** [review]

File: types/ember/index.d.ts [hunk 1]
```diff
@@ -1138,11 +1138,7 @@ declare module 'ember' {
              * function returns true for. This method corresponds to `filter()` defined in
              * JavaScript 1.6.
              */
-            filter<S extends T>(
-                callbackfn: (value: T, index: number, array: T[]) => value is S,
-                thisArg?: any
-            ): S[];
-            filter(callbackfn: (value: T, index: number, array: T[]) => any, thisArg?: any): NativeArray<T>;
```
Comment (andy-ms) [inline types/ember/index.d.ts]: This previously returned `nativeArray`, but only in one overload, was that just wrong?

File: types/ember/index.d.ts [hunk 2]
```diff
@@ -1138,11 +1138,7 @@ declare module 'ember' {
...
+            filter: GlobalArray<T>['filter'];
```
Comment (andy-ms) [inline types/ember/index.d.ts]: This used to return `NativeArray`, but only in one overload, was that just wrong?

**Review by ghost — COMMENTED** [review]
You could also do:
`interface Enumerable<T> extends Pick<GlobalArray<T>, "forEach" | "map" | "filter" | "find" | "every" | "reduce">`

File: types/ember/index.d.ts (hunk 2 above)
Comment (dwickern) [reply types/ember/index.d.ts]: `NativeArray<T>` (which extends `T[]`) would be more correct but I'm not sure we can use mapped types to convert the return type.

Comment (andy-ms) [reply types/ember/index.d.ts]: You can't without Microsoft/TypeScript#6606

**Comment by dwickern** [comment]
Returning `T[]` instead of `NativeArray<T>` we'll lose some APIs in addons (when `EXTEND_PROTOTYPES === false`)

For example this `mapBy` will no longer compile even though the method exists:
```js
> typeof Ember.ArrayProxy.create().map(x => x).mapBy
"function"
```

But I think this is a minor consequence and we need to fix these definition conflicts which affect _all_ our users. Then we can fix typings for addons if/when Microsoft/TypeScript#6606 lands

**Comment by chriskrycho** [comment]
Strongly concur with @dwickern. Also, I believe we *should* be able to recommend type-casting as a workaround here for those scenarios (or, as a last resort, `// @ts-ignore`).

**Comment by andy-ms** [comment]
OK, awaiting review from other authors.

**Comment by chriskrycho** [comment]
@andy-ms – why do you need further reviews? @dwickern wrote a ton of these type definitions in the first place, with significant contributions from me and a few others. Can you go ahead and merge please?

**Comment by andy-ms** [comment]
OK, I just saw that four authors existed and didn't know what their levels of contribution were.

**Comment by chriskrycho** [comment]
Thanks!

**Pull request merged**